internal/validation/apiextensions/v1/xrd/handler.go

```diff
@@ -0,0 +1,185 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package xrd contains internal logic linked to the validation of the v1.CompositeResourceDefinition type.
+package xrd
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	apiextv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"github.com/crossplane/crossplane-runtime/pkg/controller"
+	xperrors "github.com/crossplane/crossplane-runtime/pkg/errors"
+
+	v1 "github.com/crossplane/crossplane/apis/apiextensions/v1"
+	"github.com/crossplane/crossplane/internal/xcrd"
+)
+
+// Error strings.
+const (
+	errNotCompositeResourceDefinition = "supplied object was not a CompositeResourceDefinition"
+
+	errUnexpectedType = "unexpected type"
+)
+
+// SetupWebhookWithManager sets up the webhook with the manager.
+func SetupWebhookWithManager(mgr ctrl.Manager, _ controller.Options) error {
+	v := &validator{client: mgr.GetClient()}
+	return ctrl.NewWebhookManagedBy(mgr).
+		WithValidator(v).
+		For(&v1.CompositeResourceDefinition{}).
+		Complete()
+}
+
+type validator struct {
+	client client.Client
+}
+
+func getAllCRDsForXRD(in *v1.CompositeResourceDefinition) (out []*apiextv1.CustomResourceDefinition, err error) {
+	crd, err := xcrd.ForCompositeResource(in)
+	if err != nil {
+		return out, xperrors.Wrap(err, "cannot get CRD for Composite Resource")
+	}
+	out = append(out, crd)
+	// if claim enabled, validate claim CRD
+	if in.Spec.ClaimNames == nil {
+		return out, nil
+	}
+	crdClaim, err := xcrd.ForCompositeResourceClaim(in)
+	if err != nil {
+		return out, xperrors.Wrap(err, "cannot get Claim CRD for Composite Claim")
+	}
+	out = append(out, crdClaim)
+	return out, nil
+}
+
+// ValidateCreate validates a Composition.
+func (v *validator) ValidateCreate(ctx context.Context, obj runtime.Object) (warns admission.Warnings, err error) {
+	in, ok := obj.(*v1.CompositeResourceDefinition)
+	if !ok {
+		return nil, errors.New(errNotCompositeResourceDefinition)
+	}
+	validationWarns, validationErr := in.Validate()
+	warns = append(warns, validationWarns...)
+	if validationErr != nil {
+		return validationWarns, validationErr.ToAggregate()
+	}
+	crds, err := getAllCRDsForXRD(in)
+	if err != nil {
+		return warns, xperrors.Wrap(err, "cannot get CRDs for CompositeResourceDefinition")
+	}
+	for _, crd := range crds {
+		// Can't use validation.ValidateCustomResourceDefinition because it leads to dependency errors,
+		// see https://github.com/kubernetes/apiextensions-apiserver/issues/59
+		// if errs := validation.ValidateCustomResourceDefinition(ctx, crd); len(errs) != 0 {
+		//	return warns, errors.Wrap(errs.ToAggregate(), "invalid CRD generated for CompositeResourceDefinition")
+		//}
```

Comment:
Just wondering, do we expect this to change in the future, I see the ticket is closed as "Not Planned". Or is this note just informative?

Reply:
Just informative, we might think of reimplementing the validation logic on XRDs directly in the future, at least we have a reference.